impl/graphsync.go

```diff
@@ -470,8 +470,17 @@ func (gsr *graphSyncReceiver) ReceiveMessage(
 	ctx context.Context,
 	sender peer.ID,
 	incoming gsmsg.GraphSyncMessage) {
-	gsr.graphSync().responseManager.ProcessRequests(ctx, sender, incoming.Requests())
-	gsr.graphSync().requestManager.ProcessResponses(sender, incoming.Responses(), incoming.Blocks())
+
+	requests := incoming.Requests()
+	responses := incoming.Responses()
+	blocks := incoming.Blocks()
+
+	if len(requests) > 0 {
+		gsr.graphSync().responseManager.ProcessRequests(ctx, sender, incoming.Requests())
+	}
+	if len(responses) > 0 || len(blocks) > 0 {
```

Comment:
@hannahhoward could this possibly be made exclusive, i.e. an `else if`, or an assertion and error if there are requests then there shouldn't be responses or blocks?

It'd be nice to plumb a trace through this function but the "end" gets even tricker to determine when it forks off in these two directions, potentially for the same trace.

Reply:
I don't think it can be an else. The protocol message format supports both in the same request.

Reply:
because I'm a hack? it's done properly on my original tracing branch where I pulled this out of, not sure why it wasn't reflected here.